### PR TITLE
Fix issue 14724 - std.getopt: config.required breaks --help

### PR DESCRIPTION
**Pull request by burner**
https://issues.dlang.org/show_bug.cgi?id=14724

### CODE REVIEW

File: std/getopt.d [hunk 1]
```diff
@@ -563,14 +563,20 @@ private void getoptImpl(T...)(ref string[] args, ref configuration cfg,
...
             {
-                throw new GetOptException("Required option " ~ option ~
-                    " was not supplied");
+                requiredNotPassed = new GetOptException("Required option "
+                    ~ option ~ " was not supplied");
             }
             cfg.required = false;
 
-            return getoptImpl(args, cfg, rslt, opts[lowSliceIdx .. $]);
+            getoptImpl(args, cfg, rslt, opts[lowSliceIdx .. $]);
```
Comment (schveiguy) [inline std/getopt.d]: Doesn't this kill the tail-call optimization?

**Comment by schveiguy** [comment]
I think this can be handled easier by processing the --help option first (possibly outside the Impl function). Then you know while processing each parameter that --help has been passed or not.

**Comment by burner** [comment]
if you handle help outside impl, the "not so new h|help" functionally would break legacy code, as currently the user can use h and help for anything they want. Changing that is not good IMO. Even if that would re-enable tail-call optimization, if it was present in the first place.

**Comment by schveiguy** [comment]
This is possible to work around:
1. check to see if any --help option exist in the `options` array (shouldn't be hard, search for "help|h" string)
2. Do this:
```D
if(helpIncluded)
   return getOptImpl(args, cfg, rslt, options);
else
   return getOptImpl(args, cfg, rslt, "help|h", "This help information.", () {rslt.helpWanted = true;}, options);
```

And you can avoid all the specific help code in the implementation.

This should be identical to current implementation result, but without triggering exception on --help. As it is, I think the way this is done is terrible. What if the app isn't written in English?

**Comment by burner** [comment]
the searching part would require to iterate the complete input, properly recursive as it is done currently. I think the cure is worse than the disease (IMO it is not even a problem)

**Comment by schveiguy** [comment]
> the searching part would require to iterate the complete input, properly recursive as it is done currently.

No.

```D
foreach(opt; options)
   static if(is(typeof(opt) == string))
      if(opt == "help|h") helpIncluded = true;
```

**Comment by schveiguy** [comment]
Also note, the current code will add a "help" option at the end for displaying help, even if you have handled help on your own.

**Comment by burner** [comment]
actually I think that is not the case the in-build help is only used if there are no more options to process but you still have arguments left.

**Comment by schveiguy** [comment]
> actually I think that is not the case

```D
import std.getopt;

void main(string[] args)
{
    bool help;
    auto rslt = getopt(args, "help|h", "test help", &help);

    if (help)
    {
        defaultGetoptPrinter("Some information about the program.",
            rslt.options);
    }
}
```

```
$ ./testgetopt -h
Some information about the program.
-h --help test help
-h --help This help information.
```

**Comment by burner** [comment]
hm, my bad

**Comment by schveiguy** [comment]
Thinking about this some more. I really am kind of not liking that we handle unspecified required arguments with an exception. Even if we fix this, someone who handles --help on their own will be subject to the same issue.

I don't know a good way to fix this.

File: std/getopt.d (hunk 1 above)
Comment (DmitryOlshansky) [inline std/getopt.d]: I don't see getopt being enough of bottleneck anywhere to care about micro-optimizations like tail-call here.

Comment (schveiguy) [inline std/getopt.d]: Probably not. I'm not a huge fan of allocating the exceptions without throwing them though.

**Comment by quickfur** [comment]
Updated PR title to be more descriptive.

**Comment by burner** [comment]
@quickfur thanks

**Comment by burner** [comment]
@schveiguy @quickfur  please have a look the tail recursion returned

**Comment by quickfur** [comment]
LGTM.

**Comment by schveiguy** [comment]
I'll merge, but I found another problem that will make this problem obsolete :) See https://issues.dlang.org/show_bug.cgi?id=14921

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Pull request merged**